### PR TITLE
DI-894 development

### PR DESCRIPTION
**Pull request by jethror1**
- initial development of tool for being able to gather up all previously run dias reports jobs and rerun with latest / specified assay config file
- limited unit tests written (for now)

- - -
This change is [Reviewable](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1)

### CODE REVIEW

**Review by kjwinfield — CHANGES_REQUESTED** [review]
> Reviewed 8 of 8 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1)* status: all files reviewed, 9 unresolved discussions (waiting on @jethror1)

___
*[`bin/run_reports.py` line 40 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-cgL70SdSMDWg7f6s:-Nu-cgL70SdSMDWg7f6t:b-26lrhc) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/run_reports.py#L40)):*
> ```Python
>     ----------
>     report_jobs : list
>         list od dicts of describe details for each xlsx report
> ```

typo 'of'
___
*[`bin/utils/dx_manage.py` line 56 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-egNbAI7EvuPw1guH:-Nu-egNbAI7EvuPw1guI:b-dbihni) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/utils/dx_manage.py#L56)):*
> ```Python
>         job = selected_jobs.get(project)
> 
>         print(f"Using specified CNV job: {job}")
> ```

maybe reword to explain
e.g.
project <job> is in list of projects with multiple CNV calling jobs. Using job <job> which has been specified as the correct job.
___
*[`bin/utils/dx_manage.py` line 141 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-hDcYB8XCpSBlN4rI:-Nu-hDcYB8XCpSBlN4rJ:bhfk68v) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/utils/dx_manage.py#L141)):*
> ```Python
>             name=f"002_*{assay}",
>             name_mode="glob",
>             created_after=f"-{start}d",
> ```

maybe there should be a created after option - if the new workbooks are live on an 002 project there's no point in recreating them if theyve already been made
___
*[`bin/utils/dx_manage.py` line 217 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-husR7ywUwyvndwPd:-Nu-husS-hqhlxoRppcu:b-tf223a) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/utils/dx_manage.py#L217)):*
> ```Python
>             job_details["describe"]["output"]["xlsx_report"]["$dnanexus_link"]
>         )
>         print(job_details["describe"]["runInput"])
> ```

should possibly record somewhere (print?) samples that have no code / code could not been found
___
*[`bin/utils/dx_manage.py` line 239 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-i_C19yxkK9sAEOIt:-Nu-i_C19yxkK9sAEOIu:b-g3v97d) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/utils/dx_manage.py#L239)):*
> ```Python
>     # have manually selected one
>     # TODO - add this to config or something
>     single_dirs = {
> ```

this project also has two:
project-Ggyb2G84zJ4363x2JqfGgb6J and the correct one is /output/CEN-240322_0936

i agree this should be a config or other input file
especially because there's a two GIAB situation with CEN meaning that there will be multiple of these for a while
an alternative is to check if there are reports folders inside the directory as the GIAB only ones won't have reports, but if the multiple CNV jobs need to be in a config anyway maybe config is the way to go
___
*[`bin/utils/dx_manage.py` line 246 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-gLPKA3WszY1gwwFA:-Nu-gLPKA3WszY1gwwFB:b-ufktef) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/utils/dx_manage.py#L246)):*
> ```Python
>         path = f"{project}:{single_dirs.get(project)}"
> 
>         print(f"Using specified Dias single path: {path}")
> ```

Same as above to make it clear what this means
___
*[`bin/utils/dx_manage.py` line 261 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-jCqsBJyookki7SKV:-Nu-jCqsBJyookki7SKW:b-u5fung) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/utils/dx_manage.py#L261)):*
> ```Python
>     if len(files) > 1:
>         # TODO handle which to choose, should just be one so far
>         print(f"More than single path found, multiqc files found")
> ```

think this was probably supposed to say "More than one"? not sure what it means
___
*[`bin/utils/utils.py` line 30 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-jU5i0zz4zhghegSh:-Nu-jU5j1reUjrNmRw_7:bfld71d) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/utils/utils.py#L30)):*
> ```Python
>     results = []
> 
>     with concurrent.futures.ThreadPoolExecutor(max_workers=32) as executor:
> ```

i could be wrong but does this mean it can only be run on a 32 core instance?
___
*[`tests/test_utils.py` line 29 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-jpkOCoFCeaR6_Ohz:-Nu-jpkOCoFCeaR6_Oi-:b-gqkv9i) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/tests/test_utils.py#L29)):*
> ```Python
>         calculated_days = date_to_datetime("230302")
> 
>         assert calculated_days == 367, "time delate calculated incorrectly"
> ```

"delate" typo

**Review by jethror1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1)* status: 5 of 8 files reviewed, 8 unresolved discussions (waiting on @kjwinfield)

___
*[`bin/utils/dx_manage.py` line 56 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-egNbAI7EvuPw1guH:-Nu0Fspv1P76jrbpND_n:b-896fix) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/utils/dx_manage.py#L56)):*
<details><summary><i>Previously, kjwinfield (Katherine Winfield) wrote…</i></summary><blockquote>

maybe reword to explain
e.g.
project <job> is in list of projects with multiple CNV calling jobs. Using job <job> which has been specified as the correct job.
</blockquote></details>

Done.
___
*[`bin/utils/dx_manage.py` line 141 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-hDcYB8XCpSBlN4rI:-Nu0Fy88BoSOV1LXQkcu:b-2eq33i) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/utils/dx_manage.py#L141)):*
<details><summary><i>Previously, kjwinfield (Katherine Winfield) wrote…</i></summary><blockquote>

maybe there should be a created after option - if the new workbooks are live on an 002 project there's no point in recreating them if theyve already been made
</blockquote></details>

hmm true, we could add it as another argument, will have a think and come back to this 
___
*[`bin/utils/dx_manage.py` line 217 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-husR7ywUwyvndwPd:-Nu0GutT9y123PbpUjP_:br0ftpa) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/utils/dx_manage.py#L217)):*
<details><summary><i>Previously, kjwinfield (Katherine Winfield) wrote…</i></summary><blockquote>

should possibly record somewhere (print?) samples that have no code / code could not been found
</blockquote></details>

since it's coming from a previous job input it _should_ always exist, I think this needs some thought in terms of implementation process on what we want to log / what we would do with it, think could do with making a list of questions to discuss - will add to the list of things to think about
___
*[`bin/utils/dx_manage.py` line 239 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-i_C19yxkK9sAEOIt:-Nu0HL-E8RtTPuLiXtdd:b6vwun4) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/utils/dx_manage.py#L239)):*
<details><summary><i>Previously, kjwinfield (Katherine Winfield) wrote…</i></summary><blockquote>

this project also has two:
project-Ggyb2G84zJ4363x2JqfGgb6J and the correct one is /output/CEN-240322_0936

i agree this should be a config or other input file
especially because there's a two GIAB situation with CEN meaning that there will be multiple of these for a while
an alternative is to check if there are reports folders inside the directory as the GIAB only ones won't have reports, but if the multiple CNV jobs need to be in a config anyway maybe config is the way to go
</blockquote></details>

both of these sound good, can just select the one that has reports, and if more than one or none are found then log it and skip to review, maybe need to make a list of 'to manually review' cases that go into a review log on a sample / project level. Have added that project and path for now
___
*[`bin/utils/dx_manage.py` line 246 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-gLPKA3WszY1gwwFA:-Nu0HmJK03HEqkP0FIsi:b-896fix) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/utils/dx_manage.py#L246)):*
<details><summary><i>Previously, kjwinfield (Katherine Winfield) wrote…</i></summary><blockquote>

Same as above to make it clear what this means
</blockquote></details>

Done.
___
*[`bin/utils/dx_manage.py` line 261 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-jCqsBJyookki7SKV:-Nu0I3hZ02DbHoi6D0Uy:blrokco) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/utils/dx_manage.py#L261)):*
<details><summary><i>Previously, kjwinfield (Katherine Winfield) wrote…</i></summary><blockquote>

think this was probably supposed to say "More than one"? not sure what it means
</blockquote></details>

yep I just can't do basic English
___
*[`bin/utils/utils.py` line 30 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-jU5i0zz4zhghegSh:-Nu0I7QG6kgfpenSToA8:b7coz2x) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/utils/utils.py#L30)):*
<details><summary><i>Previously, kjwinfield (Katherine Winfield) wrote…</i></summary><blockquote>

i could be wrong but does this mean it can only be run on a 32 core instance?
</blockquote></details>

this controls the concurrent quries being made (i.e describing the reports job) to do the querying in parallel, this is before launching anything in DNAnexus. I chose 32 as somewhere in Slack I posted a thing before of timing and beyond 32 it doesn't really get any faster
___
*[`tests/test_utils.py` line 29 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-jpkOCoFCeaR6_Ohz:-Nu0IUoY1brNWFjkUh-6:b-896fix) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/tests/test_utils.py#L29)):*
<details><summary><i>Previously, kjwinfield (Katherine Winfield) wrote…</i></summary><blockquote>

"delate" typo
</blockquote></details>

Done.

**Review by jethror1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1)* status: 4 of 8 files reviewed, 8 unresolved discussions (waiting on @kjwinfield)

___
*[`bin/utils/dx_manage.py` line 141 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-hDcYB8XCpSBlN4rI:-NwA3-1m3uHKjwpVCi6V:bt27bom) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/utils/dx_manage.py#L141)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

hmm true, we could add it as another argument, will have a think and come back to this
</blockquote></details>

I think all of this is going to get scrapped in favour of directly querying Clarity / the export, so going to leave it for now
___
*[`bin/utils/dx_manage.py` line 217 at r1](https://reviewable.io/reviews/eastgenomics/dias_reports_bulk_reanalysis/1#-Nu-husR7ywUwyvndwPd:-NwA37wZ1eAH-q7b7e9I:bocdjwp) ([raw file](https://github.com/eastgenomics/dias_reports_bulk_reanalysis/blob/15927bb7ffcf52f61ca8361f5b7744c8990e1caf/bin/utils/dx_manage.py#L217)):*
<details><summary><i>Previously, jethror1 (Jethro Rainford) wrote…</i></summary><blockquote>

since it's coming from a previous job input it _should_ always exist, I think this needs some thought in terms of implementation process on what we want to log / what we would do with it, think could do with making a list of questions to discuss - will add to the list of things to think about
</blockquote></details>

same again, don't think this will be needed
